libtrap/src/trap.c

```diff
@@ -2690,17 +2594,38 @@ int trap_ctx_get_last_error(trap_ctx_t *ctx)
 
 const char *trap_ctx_get_last_error_msg(trap_ctx_t *ctx)
 {
+   /* TODO: this is dangerous... the value in buffer can be changing constantly. */
    trap_ctx_priv_t *c = ctx;
    return c->trap_last_error_msg;
 }
 
 void trap_ctx_send_flush(trap_ctx_t *ctx, uint32_t ifc)
 {
    trap_ctx_priv_t *c = ctx;
-   if (!c || !c->initialized) {
+
+   if (c == NULL || c->initialized == 0) {
       return;
    }
-   trap_store_into_buffer(c, ifc, (void *) c, 0, c->out_ifc_list[ifc].datatimeout, 1);
+
+   trap_output_ifc_t *oifc = &c->out_ifc_list[ifc];
+   if (pthread_mutex_trylock(&oifc->ifc_mtx) == 0) {
```

Comment:
I don't think you should just do nothing when the IFC is locked. It might be locked just because a message is being written into buffer, which doesn't automatically mean, that the buffer will be send. So, it may easily happen, that the request for immediate flush is ignored.
I think you should block and wait for the lock. Reasons:
* if it's locked just because something is being written to buffer, it will be unlocked very soon
* if a buffer is being sent than:
   * if timeout = NO_WAIT, the lock will be released very soon in any case
   * if timeout = WAIT (or some time), the lock can be locked for a long time, so the call of this function will block - but that's expected, since it's a request to flush buffer on an interface with blocking timeout.
(I hadn't analyzed it in detail though, I'm not sure if there aren't other places where the interface lock is being used)

Reply:
I agree. The intention was not to block when flushing the buffer.